src/main/java/org/apache/datasketches/sampling/EbppsItemsSample.java

```diff
@@ -0,0 +1,301 @@
+        data_ = new ArrayList<T>(1);
+        data_.add(item);  
+      }
+      partialItem_ = null;
+    } else {
+      data_ = null;
+      partialItem_ = item;
+    }
+  }
+
+  public void reset() {
+    c_ = 0.0;
+    partialItem_ = null;
+    data_.clear();
+  }
+
+  public ArrayList<T> getSample() {
+    final double cFrac = c_ % 1;
+    final boolean includePartial = partialItem_ != null && rand_.nextDouble() < cFrac;
+    final int resultSize = (data_ != null ? data_.size() : 0) + (includePartial ? 1 : 0);
+
+    if (resultSize == 0)
+      return null;
+
+    ArrayList<T> result = new ArrayList<>(resultSize);
+    if (data_ != null)
+      result.addAll(data_);
+
+    if (includePartial)
+      result.add(partialItem_);
+
+    return result;      
+  }
+
+  @SuppressWarnings("unchecked")
+  T[] getAllSamples(final Class<?> clazz) {
+    // Is it faster to use sublist and append 1?
+    final T[] itemsArray = (T[]) Array.newInstance(clazz, getNumRetainedItems());
+    int i = 0;
+    if (data_ != null) {
+      for (T item : data_) {
+        if (item != null) {
+          itemsArray[i++] = item;
+        }
+      }
+    }
+    if (partialItem_ != null)
+      itemsArray[i] = partialItem_; // no need to increment i again
+
+    return itemsArray;
+  }
+
+  // package-private for use in merge and serialization
+  ArrayList<T> getFullItems() {
+    return data_;
+  }
+
+  // package-private for use in merge and serialization
+  T getPartialItem() {
+    return partialItem_;
+  }
+
+  public double getC() { return c_; }
+
+  boolean hasPartialItem() { return partialItem_ != null; }
+
+  // for testing to allow setting the seed
+  void replaceRandom(Random r) {
+    rand_ = r;
+  }
+
+  public void downsample(final double theta) {
+    if (theta >= 1.0) return;
+
+    double newC = theta * c_;
+    double newCInt = Math.floor(newC);
+    double newCFrac = newC % 1;
+    double cInt = Math.floor(c_);
+    double cFrac = c_ % 1;
+
+    if (newCInt == 0.0) {
+      // no full items retained
+      if (rand_.nextDouble() > (cFrac / c_)) {
+        swapWithPartialItem();
+      }
+      data_.clear();
+    } else if (newCInt == cInt) {
+      // no items deleted
+      if (rand_.nextDouble() > (1 - theta * cFrac)/(1 - newCFrac)) {
+        swapWithPartialItem();
+      }
+    } else {
+      if (rand_.nextDouble() < theta * cFrac) {
+        // subsample data in random order; last item is partial
+        // create sample size newC then swapWithPartialItem()
+        subsample((int) newCInt);
+        swapWithPartialItem();
+      } else {
+        // create sample size newCInt + 1 then moveOneToPartialItem()
+        subsample((int) newCInt + 1);
+        moveOneToPartialItem();
+      }
+    }
+
+    if (newC == newCInt)
+      partialItem_ = null;
+
+    c_ = newC;
+  }
+
+  public void merge(final EbppsItemsSample<T> other) {
+    //double cInt = Math.floor(c_);
+    double cFrac = c_ % 1;
+    double otherCFrac = other.c_ % 1;
```

Comment:
Java modulo can produce negative numbers -- does that fact affect the following logic?

Reply:
C is the expected sample size so it better not go negative!

Reply:
Is there a check somewhere that makes sure that C can never be negative?

Reply:
Again, C must never go negative. I think the key is needing to check for non-negativity elsewhere which will prevent it from being negative here.

If someone modifies the value via reflection or something, yes, bad things will happen.

Reply:
Yes.